internal/cmd/envoy/shutdown_manager.go

```diff
@@ -0,0 +1,234 @@
+			logger.Error(err, "server shutdown error")
+		}
+		close(c)
+	}()
+
+	// Start HTTP server
+	logger.Info("starting shutdown manager")
+	if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+		logger.Error(err, "starting shutdown manager failed")
+	}
+
+	// Wait until done
+	<-c
+	return nil
+}
+
+// shutdownReadyHandler handles the endpoint used by a preStop hook on the Envoy
+// container to block until ready to terminate. When the graceful drain process
+// has begun a file will be written to indicate shutdown is in progress. This
+// handler will poll for the file and block until it is removed.
+func shutdownReadyHandler(w http.ResponseWriter, readyTimeout time.Duration, activeFile string) {
+	var startTime = time.Now()
+
+	logger.Info("received shutdown ready request")
+
+	// Since we cannot establish order between preStop hooks, wait a bit giving the exec
+	// preStop hook a chance to touch the signaling file before starting to poll for it.
+	time.Sleep(2 * time.Second)
+
+	// Exit early if active file failed to show up
+	if _, err := os.Stat(activeFile); os.IsNotExist(err) {
+		logger.Info("shutdown active file not found")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Poll for shutdown readiness
+	for {
+		// Check if ready timeout is exceeded
+		elapsedTime := time.Since(startTime)
+		if elapsedTime > readyTimeout {
+			logger.Info("shutdown readiness timeout exceeded")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		_, err := os.Stat(activeFile)
+		switch {
+		case os.IsNotExist(err):
+			logger.Info("shutdown readiness detected")
+			return
+		case err != nil:
+			logger.Error(err, "error checking for shutdown readiness")
+		case err == nil:
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Shutdown is called from a preStop hook on the shutdown-manager container where
+// it will initiate a graceful drain sequence on the Envoy proxy and block until
+// connections are drained or a timeout is exceeded.
+func Shutdown(drainTimeout time.Duration, minDrainDuration time.Duration, exitAtConnections int) error {
+	var startTime = time.Now()
+	var allowedToExit = false
+
+	// Reconfigure logger to write to stdout of main process if running in Kubernetes
+	if _, k8s := os.LookupEnv("KUBERNETES_SERVICE_HOST"); k8s && os.Getpid() != 1 {
+		logger = logging.FileLogger("/proc/1/fd/1", "shutdown-manager", v1alpha1.LogLevelInfo)
+	}
+
+	// Signal to shutdownReadyHandler that drain process is started
+	if _, err := os.Create(ShutdownActiveFile); err != nil {
+		logger.Error(err, "error creating shutdown active file")
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("initiating graceful drain with %.0f second minimum drain period and %.0f second timeout",
+		minDrainDuration.Seconds(), drainTimeout.Seconds()))
+
+	// Send request to Envoy admin API to initiate the graceful drain sequence
+	if resp, err := http.Post(fmt.Sprintf("http://%s:%d/drain_listeners?graceful&skip_exit",
```

Comment:
will this cause the health check defined in https://github.com/envoyproxy/gateway/blob/338515a6f963212bf1e20a38574fa0cc4f6aee30/internal/xds/bootstrap/bootstrap.yaml.tpl#L93 to fail ?

also curious why is `skip_exit` needed ?

Reply:
The healthcheck endpoint will not fail. It is unimportant though as the the ready status of the endpoint will immediately become false when the pod enters the terminating status triggering anything that uses the endpoint to begin deprogramming.

`skip_exit` is useful to ensure new connections are not unexpectedly blocked if a drain takes than `drain-time-s` on the Envoy proxy, so it becomes important where long drain timeouts are configured. It's a knob recently added to the drain API to support Istio's graceful drain process as they transition to native sidecars: https://github.com/envoyproxy/envoy/pull/30003 and I thought it made sense to include it so we don't have to also configure `drain-time-s` on Envoy where longer than 10m drain periods are needed.

Reply:
I think its important to fail the healthcheck endpoint, so downstream LBs can deprogram the endpoint faster

Reply:
referring to calling this endpoint explicitly as the first step of the shutdown process
https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--healthcheck-fail
cc @guydc

Reply:
I think that the LB concern is related to a specific deployment type (1 proxy per Node with `ExternalTrafficPolicy: Local`). This is already handled in K8s 1.26 by: https://kubernetes.io/blog/2022/12/30/advancements-in-kubernetes-traffic-engineering/#traffic-loss-from-load-balancers-during-rolling-updates